.github/workflows/shellcheck-debian-scripts.yml

```
@@ -0,0 +1,117 @@
+# already handle filtering the file paths for specific globs.
+# Also, we need the action to always be triggered (and succeed) if we decide to have this
+# check to be blocking the PR. Otherwise (if on.(pull_request|push).paths are defined and the check is 
+# blocking the PR), the PRs which do not have changes in these specific files will be blocked.
+# Another minor reason for not using on.(pull_request|push).paths : it does not accept environment 
+# variables so we would need to hardcode the yaml array to match env.GLOBS_TO_SHELLCHECK at the top of the file.
+on:
+  pull_request:
+    # There is a known bug in GithUb but it will most probably not affect out use case
+    # https://github.com/orgs/community/discussions/118623#discussioncomment-9087833
+    # When there are 2 PRs using the same source branch (actually the same head SHA to be more specific), with the base branch in one PR matching
+    # on.pull_request.branches and thee base branch in the second PR not matching this key,
+    # then the second PR will show these checks that were triggered by the first PR but not the second PR.
+    branches:
+      - 'ubuntu/**'
+  push:
+    branches:
+      - 'ubuntu/**'
+
+concurrency:
+  group: 'ci-${{ github.workflow }}-${{ github.ref }}'
+  cancel-in-progress: true
+
+defaults:
+  run:
+    # Using bash instead of the usual sh in the other workflows in order to support the globstar **
+    shell: bash -ex {0}
+
+# Note: No need to specify the shell option in the shellcheck command.
+# These linted scripts are for ubuntu so the !/bin/sh (i.e. system shell) shebang at the top of these scripts .
+# So no need to pass the shell option to the shellcheck command as shellcheck will read the shebang at the top of the file (sh) so will
+# use sh
+jobs:
+  shellcheck-on-all-matching-files:
+    name: ShellCheck on all matching files in the repo
+    runs-on: ubuntu-24.04
+    continue-on-error: true
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v4
+
+      - name: Run ShellCheck
+        run: |
+          shopt -s globstar
+          echo "The globs to check are ${GLOBS_TO_SHELLCHECK}"
+          ALL_MATCHING_SCRIPTS=$(git ls-files ${GLOBS_TO_SHELLCHECK} | tr '\n' ' ')
+          echo "The matching scripts to shellcheck are ${ALL_MATCHING_SCRIPTS}"
+          if [ -z "${ALL_MATCHING_SCRIPTS}" ]
+          then
+            echo "There are no files in the repo which match the glob patterns so shellcheck will not run"
+          else
+            shellcheck ${ALL_MATCHING_SCRIPTS}
+          fi
+
+  shellcheck-on-matching-and-changed-files:
+    name: ShellCheck on matching files that have changed
+    runs-on: ubuntu-24.04
+    continue-on-error: true
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v4
+
+      - name: Get all matching changed files
+        id: matching-changed-files
+        uses: tj-actions/changed-files@v46
```

While the workflow approach looks good I'm hesitant to adopt the use of tj-actions to get changed-file lists given the recent supply-chain attack CVE https://www.wiz.io/blog/github-action-tj-actions-changed-files-supply-chain-attack-cve-2025-30066.  I think we'd probably prefer just using git command line direct with something like the following instead:
```
              git diff --name-only ${{ github.event.pull_request.base.sha }} ${{ github.sha }}
```

Given the severity of the CVE potentially exploiting github secrets, I'd rather cloud-init upstream project avoid dependency on this workflow tool if we can solve this ourselves in the absence of tj-actions.

---

Another alternative could be using a specific SHA1 hash instead of a tag name. v46.

---

Our security team agreed with the approach of [pinning known good SHAs](https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/find-and-customize-actions#using-shas)   So, let's set this to the SHA relative to v47
```suggestion
        uses: tj-actions/changed-files@1ac60965030588f3b867cdd4a3900f37ec99970c
```

---

I think the v47 is represented by this SHA https://github.com/tj-actions/changed-files/commit/24d32ffd492484c1d75e0c0b894501ddb9d30d62 so I used it